src/key_gen/mod.rs

```diff
@@ -0,0 +1,930 @@
+    // agreement reached, and the message should be multicast to all nodes.
+    fn handle_initialization<R: RngCore>(
+        &mut self,
+        sec_key: &S,
+        rng: &mut R,
+        m: usize,
+        n: usize,
+        sender: u64,
+        member_list: BTreeSet<S::PublicId>,
+    ) -> Result<Vec<Message<S::PublicId>>, Error> {
+        if self.phase != Phase::Initialization {
+            return Err(Error::Unknown);
+        }
+
+        if let Some((m, _n, member_list)) =
+            self.initalization_accumulator
+                .add_initialization(m, n, sender, member_list)
+        {
+            self.threshold = m;
+            self.pub_keys = member_list;
+            self.phase = Phase::Contribution;
+
+            let mut rng = rng_adapter::RngAdapter(&mut *rng);
+            let our_part = BivarPoly::random(self.threshold, &mut rng);
+            let commitment = our_part.commitment();
+            let encrypt = |(i, pk): (usize, &S::PublicId)| {
+                let row = our_part.row(i + 1);
+                sec_key
+                    .encrypt(pk, &serialize(&row)?)
+                    .ok_or(Error::Encryption)
+            };
+            let rows = self
+                .pub_keys
+                .iter()
+                .enumerate()
+                .map(encrypt)
+                .collect::<Result<Vec<_>, Error>>()?;
+
+            let mut result = Vec::new();
+            result.push(Message::Contribution {
+                key_gen_id: self.our_index,
+                part: Part(commitment, rows),
+            });
+            return Ok(result);
+        }
+        Ok(Vec::new())
+    }
+
+    // Handles a `Contribution` message.
+    // If it is invalid, sends a `Complaint` message targeting the sender to be broadcast.
+    // If all contributed, returns a `Commitment` message to be broadcast.
+    fn handle_contribution<R: RngCore>(
+        &mut self,
+        sec_key: &S,
+        rng: &mut R,
+        sender_index: u64,
+        part: Part,
+    ) -> Result<Vec<Message<S::PublicId>>, Error> {
+        if self.phase != Phase::Contribution {
+            return Err(Error::Unknown);
+        }
+
+        let sender_id = self
+            .node_id_from_index(sender_index)
+            .ok_or(Error::UnknownSender)?;
+
+        if self
+            .contribution_accumulator
+            .add_contribution(sender_id, sender_index, part)
+        {
+            self.finalize_contributing_phase(sec_key, rng)
+        } else {
+            Ok(Vec::new())
+        }
+    }
+
+    // TODO: So far this function has to be called externally to indicates a completion of the
+    //       contribution phase. That is, the owner of the key_gen instance has to wait for a fixed
+    //       interval, say an expected timer of 5 minutes, to allow the messages to be exchanged.
+    //       May need to be further verified whether there is a better approach.
+    pub fn finalize_contributing_phase<R: RngCore>(
```

Comment:
I wonder if it would be possible to reduce the API surface a bit by merging the `finalize_contributing_phase` and `finalize_complaining_phase` into one function which automatically detects what to do based on the current phase. Would that make sense?

Reply:
good idea, will give a try within the next PR